api/routes.py

```diff
@@ -269,23 +270,40 @@ def put(self, task_data, task_id):
                 abort(409, message="This task has already been completed.")
             elif flask.session and flask.session["token"]:
                 tool = task_data["tool"]
-                data_obj = build_request(task_data)
-                result = toolhub_client.put(tool, data_obj)
-                if result == 200:
-                    username = get_current_user()
-                    task.user = username
-                    task.timestamp = datetime.datetime.now(datetime.timezone.utc)
-                    db.session.add(task)
-                    try:
-                        db.session.commit()
-                        return f'{task_data["field"]} successfully updated for {tool}.'
-                    except exc.DBAPIError as err:
-                        print(err)
-                        abort(503, message="Database connection failed.")
+                data = build_request(task_data)
+                token = flask.session["token"]["access_token"]
+                put_task = make_put_request.delay(tool, data, token)
+                response = put_task.get()
```

Comment:
For the task to actually happen in the background and not block the task/put route execution, whatever happens after this point related to the task has to be handled in the task, including error checking, updating our db, etc. This means that as soon as we launch the task, we need to return from the function.

Reply:
I'd been thinking about that and I'm not sure what to return, since at that moment there isn't a guarantee that the change will have been made.  So what exactly do we tell the user and when do we tell them that their attempt to update Toolhub was a success?  Or if it wasn't, how do we inform them of failure?

I admit that the more I thought about it, the more I started to doubt the value of doing this in an asynchronous way at all.

Reply:
So `make_put_request()` would look something like this:

```
@shared_task(bind=True)  # Bound tasks are needed for retries, for accessing information about the current task request, and more
def make_put_request():
    try:
        # make the put request here
    except Exception as e:
        print(f'exception {e} raised, retrying after 5 seconds....')  # this should be logging, really
        raise self.retry(exc=e, countdown=5). # dummy example, there are better ways of configuring retries
    else:
        # code that executes after a successful try block goes here
    finally:
        # code that needs to run whether the try block was successful or not goes here
```

Reply:
The above was not an answer to your question, we were commenting at the same time. :)

> I'd been thinking about that and I'm not sure what to return, since at that moment there isn't a guarantee that the change will have been made. So what exactly do we tell the user and when do we tell them that their attempt to update Toolhub was a success? Or if it wasn't, how do we inform them of failure?

I wouldn't tell them anything, tbh. The only reason a task could fail should be if Toolhub is down for a longer period. Anything else would be a bug on our side, and we'd need to be made aware of it through testing/logging. When the user navigates to their dashboard, they should be able to see the recently completed tasks.

Reply:
> I wouldn't tell them anything, tbh. The only reason a task could fail should be if Toolhub is down for a longer period. Anything else would be a bug on our side, and we'd need to be made aware of it through testing/logging. When the user navigates to their dashboard, they should be able to see the recently completed tasks.

True.  At the moment there's no input validation on the front end, but I'll push for some to be included.  I can also run validation checks on the backend just to be sure.

Reply:
With all of that in place then, no, there should never be a reason for failure.

Reply:
> I admit that the more I thought about it, the more I started to doubt the value of doing this in an asynchronous way at all.

The value of async is that potentially slow communication between our backend and Toolhub is guaranteed to not block the /task/put endpoint, especially if many users are using the application at the same time, like during a hackathon/edit-a-thon

Reply:
That makes sense.  I'm just paranoid about unexpected failures.  🤣